src/flightRecorder.cpp

```diff
@@ -655,6 +685,86 @@ class Recording {
         _last_gc_id = gc_id;
     }
 
+    // Helper function to determine if a process should be included based on thresholds
+    inline bool shouldIncludeProcess(const ProcessInfo* info) {
+        return info->_cpu_percent >= MIN_CPU_THRESHOLD || info->_vm_rss >= MIN_MEMORY_THRESHOLD_KB;
+    }
+
+    void populateCpuPercent(ProcessInfo* info) {
+        int pid = info->_pid;
+        u64 current_cpu_total = info->_cpu_user + info->_cpu_system;
+        u64 current_time = info->_last_update;
+
+        ProcessHistory& history = _process_history[pid];
+        if (history.prev_timestamp == 0) {
+            history.prev_cpu_total = current_cpu_total;
+            history.prev_timestamp = current_time;
+            return;
+        }
+
+        u64 delta_cpu = current_cpu_total - history.prev_cpu_total;
+        u64 delta_time = current_time - history.prev_timestamp;
+
+        float cpu_percent = 0.0F;
+
+        if (delta_time > 0) {
+            float cpu_secs = static_cast<float>(delta_cpu) / OS::clock_ticks_per_sec;
+            float elapsed_secs = static_cast<float>(delta_time) / 1e9;
+            if (elapsed_secs > 0) {
+              cpu_percent = (cpu_secs / elapsed_secs) * 100.0;
+            }
+        }
+
+        info->_cpu_percent = cpu_percent;
+        history.prev_cpu_total = current_cpu_total;
+        history.prev_timestamp = current_time;
+    }
+
+  void cleanupProcessHistory(const int* activePids, int pidCount) {
+      std::unordered_set<int> active_pid_set(activePids, activePids + pidCount);
+      for (auto it = _process_history.begin(); it != _process_history.end(); ) {
+          if (active_pid_set.count(it->first) == 0) {
+              it = _process_history.erase(it);
+          } else {
+              ++it;
+          }
+      }
+    }
+
+    bool processMonitorCycle(u64 wall_time) {
+        if (_process_sampling_interval < 0) return false;
+
+        if ((wall_time - _last_proc_sample_time) < (u64)_process_sampling_interval) {
+            return false;
+        }
+
+        u64 start_time = OS::nanotime();
+        int pids[MAX_PROCESSES];
+        int pid_count = OS::getProcessIds(pids, MAX_PROCESSES);
+        cleanupProcessHistory(pids, pid_count);
+
+        for (int i = 0; i < pid_count; i++) {
+            if ((OS::nanotime() - start_time) > MAX_TIME_NS) {
+                Log::debug("Incomplete process sampling cycle.");
+                break;
+            }
+
+            ProcessInfo info;
+            if (OS::getBasicProcessInfo(pids[i], &info)) {
+                info._last_update = OS::nanotime();
```

Comment:
Sometimes, when TSC is not available/reliable, `OS::nanotime()` does an expensive syscall.
Avoid redundant calls when possible; in this case, you can reuse a timestamp taken in the beginning of the loop iteration.